src/osd/PeeringState.cc

```diff
@@ -5369,6 +5369,34 @@ boost::statechart::result PeeringState::Active::react(const MNotifyRec& notevt)
     if (ps->have_unfound() || (ps->is_degraded() && ps->might_have_unfound.count(notevt.from))) {
       ps->discover_all_missing(context< PeeringMachine >().get_query_map());
     }
+
+    // find valid stray pg to join acting set iff the pg in peered state
+    if (ps->acting.size() < ps->pool.info.min_size) {
+      if (notevt.notify.info.is_incomplete() ||
+          notevt.notify.info.last_update < ps->info.log_tail) {
+        return discard_event(); 
+      }
+
+      vector<int> newacting(ps->acting);
+      if (ps->pool.info.is_replicated()) {
+        for (vector<int>::const_iterator i = ps->acting.begin();
+             i != ps->acting.end();
+             ++i) {
+          if (notevt.from.osd == *i) {
+            return discard_event();
+          }
+        }
+        newacting.push_back(notevt.from.osd);
+      } else {
+        auto &shard = ps->acting[notevt.from.shard];
+        if (shard != CRUSH_ITEM_NONE || shard == notevt.from.osd) {
+          return discard_event();
+        }
+        newacting[notevt.from.shard] = notevt.from.osd;
+      }
```

Comment:
First, I think the condition should probably be is_peered && !is_active.  Also, this probably isn't right since it's kind of duplicating choose_acting.  Can you simply transition to Reset?  peered && !active means we didn't persist les or anything else to mark this interval, so resetting the peering process ought to be safe and sufficient.  I think this situation is analogous to Peering::react(const AdvMap &advmap).

Also, as I said above, it would be much much much easier to merge a fix for this if you can add a standalone test to qa/standalone/osd (and that goes for essentially any PR) to demonstrate that your fix actually works and that we won't rebreak it in the future.

Reply:
Thanks for review, to be honest, I used try to simply transit the primary to Reset, but it's hard to transit the slave and stray pgs to Reset synchronously. current we will keep all participants reset synchronously because they have the same reaction on new AdvMap. here when we got a stray pg's notify, there is no new osdmap, so if we need to make a new osdmap affect the target pg,  and I think the pg temp may be the right approach..  make sense?

I will add a test ..

Reply:
Oh, you are correct, that's actually a problem.  I suppose MOSDPGQuery could have a flag indicating that the receiving pg should restart peering, but that seems clumsy.  Need to think about this a bit.

Reply:
Adding a test will certainly help to test solutions!

Reply:
I think the strategy needs to be to call choose_acting there.